### ISSUE TITLE
Reading multiple rows from json document

### ISSUE THREAD

**Issue by thbley** [duckdb/duckdb#6251]
### What happens?

Expected: 1 row, Actual: 1 row

```
D INSTALL 'httpfs';
D LOAD 'httpfs';
D SELECT * FROM read_json_auto('https://jsonplaceholder.typicode.com/todos/42') ;
┌────────┬────────┬─────────────────────────────────────────┬───────────┐
│ userId │   id   │                  title                  │ completed │
│ uint64 │ uint64 │                 varchar                 │  boolean  │
├────────┼────────┼─────────────────────────────────────────┼───────────┤
│      3 │     42 │ rerum perferendis error quia ut eveniet │ false     │
└────────┴────────┴─────────────────────────────────────────┴───────────┘
```

Expected: 200 rows, Actual: 0 rows
```
D SELECT * FROM read_json_auto('https://jsonplaceholder.typicode.com/todos') ;
┌────────┐
│  json  │
│ int32  │
├────────┤
│ 0 rows │
└────────┘
D 
```

### To Reproduce

```
SELECT * FROM read_json_auto('https://jsonplaceholder.typicode.com/todos') ;
```

=> 0 rows

```
curl -s https://jsonplaceholder.typicode.com/todos | jq length
200
```

### OS:

x64

### DuckDB Version:

v0.7.0

### DuckDB Client:

cli

### Full Name:

Thomas Bley

### Affiliation:

myself

### Have you tried this on the latest `master` branch?

- [X] I agree

### Have you tried the steps to reproduce? Do they include all relevant data and configuration? Does the issue you report still appear there?

- [X] I agree

**Comment by lnkuiper** [comment]
Somehow we're getting 0 bytes from the file at the URL, this also gives 0 rows:
```sql
SELECT * FROM read_csv('https://jsonplaceholder.typicode.com/todos', columns={'json': 'JSON'}, delim=NULL, header=0, quote=NULL, escape=NULL);
```

If we do this:
```sh
wget https://jsonplaceholder.typicode.com/todos
```

And then read the local file like so:
```sql
select * from read_json_auto('todos');
```

We get this:

|                              json <br /> struct(userid ubigint, id ubigint, title varchar, completed boolean)[]                              |
|---------------------------------------------------------------------------|
| [{'userId': 1, 'id': 1, 'title': delectus aut autem, 'completed': false}, ...] |

Which is a list of the 200 TODO's.

I don't know much about the httpfs extension. This could be a problem with the website or with httpfs. Maybe @samansmink can have a look?

**Comment by Mytherin** [comment]
This might be related to #5924

**Comment by thbley** [comment]
When downloading the file, I get internal error or segfault for select count(*):

```
wget -O todos.json https://jsonplaceholder.typicode.com/todos
./duckdb 
v0.7.0 f7827396d7
Enter ".help" for usage hints.
Connected to a transient in-memory database.
Use ".open FILENAME" to reopen on a persistent database.
D SELECT count(*) FROM read_json_auto('todos.json') ;
Segmentation fault
```

```
./duckdb 
v0.7.0 f7827396d7
Enter ".help" for usage hints.
Connected to a transient in-memory database.
Use ".open FILENAME" to reopen on a persistent database.
D SELECT * FROM read_json_auto('todos.json') ;
┌──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
│                                                                                                                                                           json                                                                                                                                                           │
│                                                                                                                          struct(userid ubigint, id ubigint, title varchar, completed boolean)[]                                                                                                                          │
├──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ [{'userId': 1, 'id': 1, 'title': delectus aut autem, 'completed': false}, {'userId': 1, 'id': 2, 'title': quis ut nam facilis et officia qui, 'completed': false}, {'userId': 1, 'id': 3, 'title': fugiat veniam minus, 'completed': false}, {'userId': 1, 'id': 4, 'title': et porro tempora, 'completed': true}, {'u…  │
└──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
D SELECT count(*) FROM read_json_auto('todos.json') ;
Error: INTERNAL Error: Unexpected type at JSON Transform UNDEFINED
```

**Comment by lnkuiper** [comment]
@thbley Thanks for the bug report. I am currently fixing bugs as they are coming in, and this no longer segfaults on my fork.

**Comment by samansmink** [comment]
The cause for this is the server not supporting content length nor range requests on the `https://jsonplaceholder.typicode.com/todos` url. This falsely sets the file size to 0 bytes thus causing the read to fail. 

Since we currently don't support full file downloads, ill PR an improved error message to notify the user here. The proper fix though is to have an (optional) fallback to full file downloads similarly to what duckdb-wasm does.

**Comment by thbley** [comment]
Thanks for the feedback!

A few questions:
- Is there a way to get multiple rows from "select * from read_json_auto('https://foo.../todos')" ?
- Would it be possible to add a custom http header (bearer auth, etc.)
- Create joins between REST API endpoints and fetch each endpoint in parallel?

**Comment by lnkuiper** [comment]
@thbley I can comment on your JSON question: you cannot get anything from the URL like Sam said, because the website does not support it.

However, if you download the file, it should work fine. You will get a single list containing structs if you use `read_json_auto`, but you can unnest both the list and the struct if you like:

```sql
select unnest(json).* from read_json_auto('todos.json');
```

**Comment by samansmink** [comment]
@thbley 
Adding a custom http header is not something we currently support. This could be an interesting for sure

We do have `http_timeout` setting! try `SET http_timeout=1000;` for 1s timeout.

You can certainly do queries that use multiple endpoints, for S3 this requires using the [qualified urls](https://github.com/duckdb/duckdb/pull/5774).

**Comment by lnkuiper** [comment]
Just documenting here that we can now read the `todos.json` without unnesting / exploding the struct. I've reworked the format detection (in #6319) and we can now simply do:
```sql
select * from 'todos.json';
```
And we will get a table with 200 rows.

Seeing as the HTTP issue has been fixed in #6293, this issue can be closed.
